### PR TITLE
bpo-37933: do not segfault canceling never started faulthandler dump

### PR DESCRIPTION
**Pull request by tacaswell**
If there is no lock, do not try to release it.

The issue is that https://github.com/python/cpython/pull/15358 deferred creating the lock until `faulthandler_dump_traceback_later` is called.  This now short-circuits if there is no lock (as we are sure there is nothing to cancel!).

https://bugs.python.org/issue37933

### CODE REVIEW

**Review by epicfaace — APPROVED** [review]

**Review by aeros — COMMENTED** [review]
Thanks for the PR @tacaswell. I've added the ``skip news`` label for now, but I'll remove it if the core devs think it's needed for this PR.

I would recommend adjusting the PR title to something a bit more focused on the area in which the changes are affecting, in order for core developers experienced with ``faulthandler`` to more easily be able find this PR: ``Fix faulthandler segfault``. This title would not include all of the details, but that can be read in the description and on the bpo issue page.

~~On my end, I was not able to replicate the segfault from the latest commit to python:master~~ (OS: Arch Linux 5.2.9): [Edit: See comments below]
image
That of course does not mean there's not an issue with the current behavior of ``faulthandler``'s stack allocation, just that I was personally unable to confirm the presence of the bug. 

Since Victor was the one who modified the behavior in https://github.com/python/cpython/pull/15358, I'll notify him.

/cc @vstinner

**Comment by epicfaace** [comment]
@aeros167 I was able to replicate the segfault. Looks like the Python version you are running is from this commit -- https://github.com/python/cpython/tree/3e54b57531. Can you try building the latest from master and see if you can replicate it?

**Comment by aeros167** [comment]
@epicfaace 

> Looks like the Python version you are running is from this commit -- https://github.com/python/cpython/tree/3e54b57531. Can you try building the latest from master and see if you can replicate it?

Good catch! I think I managed to end up on that commit when I used ``git reset --hard upstream/master`` to perform a clean build, and I had not properly updated my upstream remote, I had only updated origin. If I had used ``git reset --hard master`` in the first place though that would not have occurred. That might be a sign that I should go get some sleep after this. (:

Anyways, I was able to replicate the segfault from the latest commit:
image

Next, I'll verify that the PR branch fixes the segfault.

**Review by aeros — APPROVED** [review]
The PR branch resolves the segfault:

image

**Comment by aeros167** [comment]
Thanks for the fix @tacaswell! Apologies for the earlier mixup, I think I may have been trying to do a little bit too much in one night.

**Comment by tacaswell** [comment]
@aeros167 no worries, I have been there :wink:

File: Lib/test/test_faulthandler.py
Comment (vstinner) [inline Lib/test/test_faulthandler.py]: please add one indentation level to the dedent code.

File: Lib/test/test_faulthandler.py
Comment (vstinner) [inline Lib/test/test_faulthandler.py]: Please just wrote 0 rather than 0.

**Review by vstinner — CHANGES_REQUESTED** [review]

File: Lib/test/test_faulthandler.py
Comment (vstinner) [inline Lib/test/test_faulthandler.py]: Please add a comment like:

```
# bpo-37933: Calling cancel_dump_traceback_later()
# without dump_traceback_later() must not crash
```

Maybe rename the test to test_cancel_without_dump_traceback_later().

**Comment by vstinner** [comment]
> Most of the time it's not, but I've seen similar issues come up occasionally with Travis as well. If the build errors look to be unrelated to the PR, such as in this case, the best course of action is usually to just wait a bit and try to rerun the tests by adding an empty comment. Worst case scenario, someone has to repair the CI tests or a core developer can merge the PR without a test passing.

It's better to investigate every single CI failure which may hide a real bug. See https://pythondev.readthedocs.io/ci.html for more info about Python CI infra. I'm part of the Night’s Watch which opens a bug report for every single CI failure, like https://bugs.python.org/issue37959 this morning.

**Comment by tacaswell** [comment]
Is there anything else I need to do on this?

File: Lib/test/test_faulthandler.py
Comment (vstinner) [inline Lib/test/test_faulthandler.py]: There is a typo, no?

```suggestion
        # without dump_traceback_later() must not segfaust
```

Comment (vstinner) [reply Lib/test/test_faulthandler.py]: Note: I'm not allowed to merge my suggestion (push into your branch, from GitHub web UI).

**Comment by tacaswell** [comment]
The CI failure for 9b4cced was apt timing out downloading it's index files.

**Review by vstinner — APPROVED** [review]
LGTM.

**Comment by aeros167** [comment]
@vstinner 
> It's better to investigate every single CI failure which may hide a real bug.

If the CI failure looks unrelated to what the PR is modifying/affecting, should it be reported as a bug if it hasn't been already reported? Unrelated CI failures are fairly uncommon, this is probably just the third time I've seen it occur. In one of my previous PRs, the Travis CI failure was happening across multiple PRs and Ned was already working on fixing it.

Also, is the troubleshooting process of adding an empty comment and removing it to retrigger the tests generally advisable after investigating the CI failure and reporting the potential issue (if it hasn't already been reported)? From my understanding, this would at least prove the failures are likely unrelated to the PR itself, and indicate an underlying bug with the CI tests.

> I'm part of the Night’s Watch which opens a bug report for every single CI failure, like https://bugs.python.org/issue37959 this morning.

Is there a way to become involved with the Night's Watch as someone who isn't a core developer, other than reporting potential CI bugs encountered in PRs? I have a general interest in gaining more DevOps related experience, in particular with the building and testing parts (which both revolve
heavily around continuous integration).

**Comment by vstinner** [comment]
>  If the CI failure looks unrelated to what the PR is  modifying/affecting, should it be reported as a bug if it hasn't been  already reported?

Yes, open a new bug. But it's better to check if it has been already reported. For example, search the file name of the failing test (ex: search bugs with "test_os" in their title) or by the test method in a full text search.

> Unrelated CI failures are fairly uncommon, this is  probably just the third time I've seen it occur.

Maybe it's uncommon on your PRs, but it happened more times on other PRs. You can have a look at Travis CI build history:
https://travis-ci.org/python/cpython/builds

Such link can be found in https://pythondev.readthedocs.io/ci.html page.

> In one of my previous  PRs, the Travis CI failure was happening across multiple PRs and Ned was  already working on fixing it.

Opening an issue can help to coordinate efforts to fix a bug.

> Also, is the troubleshooting process of adding an empty comment and removing it to retrigger the tests generally advisable after  investigating the CI failure and reporting the potential issue (if it  hasn't already been reported)?

I dislike such practice. I prefer to go to the CI and trigger manually a new build. I think that only core developers are allowed to do that. The workaround is no core dev is available is to close/reopen the PR.

If a PR fails for an unrelated reason, you don't have to trigger a new job: wait until a core developer approve the PR, and they can fix the PR for you.

> Is there a way to become involved with the Night's Watch as someone  who isn't a core developer, other than reporting potential CI bugs  encountered in PRs?

You can join the python-buildbot mailing list, introduce yourself, and ask how to help.

In general, spotting bugs is the trivial task. The most time consuming part of the job is to investigate bugs.

Search for bugs which contains "buildbot" to "test" in their titles. Or look at bugs reported by me or Pablo, the two who report most of the buildbot failures :-) There are tons of "unstable" tests:
https://pythondev.readthedocs.io/unstable_tests.html
(I don't maintain the list of "Open Issues" of this page, it's outdated and far from being completed)

For example, multiprocessing tests are the most unstable tests.

**Comment by vstinner** [comment]
cc @pablogsal

**Comment by vstinner** [comment]
We should have such discussion on the python-buildbot mailing list rather on a closed pull request :-)

**Comment by aeros167** [comment]
@vstinner:
> We should have such discussion on the python-buildbot mailing list rather on a closed pull request :-)

Thank you very much for the detailed answers though, I greatly appreciate it. (: 

I'll definitely be subscribing to python-buildbot ML, I actually wasn't aware of the existence of it. So far I've mostly been active on python-dev and python-ideas, but I'm starting to realize there's many more mailing lists to get involved with. Admittedly, it can get a bit overwhelming at times (in a good way though).

In the next couple of days, I might be a tad preoccupied with preparing for a hurricane coming this weekend, but I will be sure to download some pages from https://pythondev.readthedocs.io/ci.html for offline reading in case we're without internet for a few days.

**Pull request merged**
